File: Soundfolio/Data/EntityNormalize.swift
import Foundation

enum EntityNormalize {
    static func isCatalogTrackId(_ trackId: String) -> Bool {
        let id = trackId.trimmingCharacters(in: .whitespacesAndNewlines)
        if id.isEmpty { return false }
        if id.hasPrefix("lfm-") && !id.hasPrefix("lfm-track-") { return false }
        return true
    }

    static func key(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .lowercased()
    }

    static func label(_ value: String) -> String {
        value
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
    }

    static func matches(_ a: String, _ b: String) -> Bool {
        key(a) == key(b)
    }

    static func betterDisplay(_ current: String, _ candidate: String) -> String {
        let left = label(current)
        let right = label(candidate)
        if left.isEmpty { return right }
        if right.isEmpty { return left }
        if key(left) != key(right) { return left }

        func score(_ value: String) -> Int {
            let titled = value.split(separator: " ").filter { word in
                guard let first = word.first else { return false }
                return first.isUppercase
            }.count
            return titled * 10 + value.count
        }

        return score(right) > score(left) ? right : left
    }

    static func trackGroupKey(trackId: String, trackName: String, artistName: String) -> String {
        if isCatalogTrackId(trackId) {
            return "id:\(trackId.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())"
        }
        return "name:\(key(trackName))\0\(key(artistName))"
    }

    static func catalogTrackId(trackId: String, trackName: String, artistName: String) -> String {
        if isCatalogTrackId(trackId) {
            return trackId.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        return trackGroupKey(trackId: "", trackName: trackName, artistName: artistName)
    }

    static func albumGroupKey(albumName: String, artistName: String) -> String {
        "\(key(albumName))\0\(key(artistName))"
    }

    static func artistGroupKey(artistName: String) -> String {
        key(artistName)
    }
}

File: lib/entity-normalize.ts
export function normalizeEntityKey(value: string): string {
  return value.trim().replace(/\s+/g, " ").toLocaleLowerCase();
}

export function cleanEntityLabel(value: string): string {
  return value.trim().replace(/\s+/g, " ");
}

/** Stable catalog id (Spotify, lfm-track-*, etc.) — not per-scrobble lfm-* stream ids. */
export function isCatalogTrackId(trackId: string): boolean {
  const id = trackId.trim();
  if (!id) return false;
  if (id.startsWith("lfm-") && !id.startsWith("lfm-track-")) return false;
  return true;
}

export function matchesEntity(a: string, b: string): boolean {
  return normalizeEntityKey(a) === normalizeEntityKey(b);
}

/** Prefer title-cased / longer spellings when merging rows for the same entity. */
export function pickBetterDisplayName(current: string, candidate: string): string {
  const left = cleanEntityLabel(current);
  const right = cleanEntityLabel(candidate);
  if (!left) return right;
  if (!right) return left;
  if (normalizeEntityKey(left) !== normalizeEntityKey(right)) return left;

  const score = (value: string) => {
    const words = value.split(" ");
    const titled = words.filter((word) => /^[A-Z]/.test(word)).length;
    return titled * 10 + value.length;
  };

  return score(right) > score(left) ? right : left;
}

export function trackGroupKey(trackId: string, trackName: string, artistName: string): string {
  if (isCatalogTrackId(trackId)) return `id:${trackId.trim().toLocaleLowerCase()}`;
  return `name:${normalizeEntityKey(trackName)}\0${normalizeEntityKey(artistName)}`;
}

export function catalogTrackId(trackId: string, trackName: string, artistName: string): string {
  if (isCatalogTrackId(trackId)) return trackId.trim();
  return trackGroupKey("", trackName, artistName);
}

export function albumGroupKey(albumName: string, artistName: string): string {
  return `${normalizeEntityKey(albumName)}\0${normalizeEntityKey(artistName)}`;
}

export function artistGroupKey(artistName: string): string {
  return normalizeEntityKey(artistName);
}
